gpcontrib/arenadata_toolkit/src/arenadata_toolkit_guc.c

```diff
@@ -0,0 +1,265 @@
+#include "arenadata_toolkit_guc.h"
+
+#include "cdb/cdbvars.h"
+#include "catalog/objectaccess.h"
+#include "catalog/pg_db_role_setting.h"
+#include <limits.h>
+#include "utils/guc.h"
+#include "tf_shmem.h"
+
+int			bloom_size = DEFAULT_BLOOM_SIZE_BYTES;
+int			db_track_count = DEFAULT_DB_TRACK_COUNT;
+bool		is_tracked = DEFAULT_IS_TRACKED;
+bool		get_full_snapshot_on_recovery = DEFAULT_GET_FULL_SNAPSHOT_ON_RECOVERY;
+int			drops_count = DEFAULT_DROPS_COUNT;
+char	   *tracked_schemas = DEFAULT_TRACKED_SCHEMAS;
+char	   *tracked_rel_storages = DEFAULT_TRACKED_REL_STORAGES;
+char	   *tracked_rel_kinds = DEFAULT_TRACKED_REL_KINDS;
+int			tracking_worker_naptime_sec = DEFAULT_NAPTIME_SEC;
+
+static bool is_tracked_unlocked = false;
+static bool is_get_full_snapshot_on_recovery_unlocked = false;
+static bool is_schemas_unlocked = false;
+static bool is_relkinds_unlocked = false;
+static bool is_relstorages_unlocked = false;
+
+void
+tf_guc_unlock_tracked_once(void)
+{
+	is_tracked_unlocked = true;
+}
+
+void
+tf_guc_unlock_full_snapshot_on_recovery_once(void)
+{
+	is_get_full_snapshot_on_recovery_unlocked = true;
+}
+
+void
+tf_guc_unlock_schemas_once(void)
+{
+	is_schemas_unlocked = true;
+}
+
+void
+tf_guc_unlock_relkinds_once(void)
+{
+	is_relkinds_unlocked = true;
+}
+
+void
+tf_guc_unlock_relstorages_once(void)
+{
+	is_relstorages_unlocked = true;
+}
+
+/*
+ * Prohibit changing the GUC value manually except several cases.
+ * This is not called for RESET, so RESET is not guarded
+ */
+static bool
+check_guc(bool *toolkit_guc, GucSource source, bool *manual)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && *toolkit_guc))
+	{
+		*toolkit_guc = false;
+
+		if (source != PGC_S_DATABASE && source != PGC_S_DEFAULT && source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	*manual = true;
+	return false;
+}
+
+/*
+ * Prohibit changing the arenadata_toolkit.tracking_is_db_tracked value manually
+ */
+static bool
+check_tracked(bool *newval, void **extra, GucSource source)
+{
+	bool manual = false;
+	if (check_guc(&is_tracked_unlocked, source, &manual))
+		return true;
+
+	if (manual)
+		GUC_check_errmsg("cannot change tracking status outside the tracking_register_db function");
+	return false;
```

Comment:
how about such logic
```suggestion
	return check_guc(&is_tracked_unlocked, source, "cannot change tracking status outside the tracking_register_db function");
```
and function check_guc can raise error

Reply:
`GUC_check_errmsg("%s", err_msg);`

does not work
-ERROR:  cannot change tracking status outside the tracking_register_relstorages function
+ERROR:  missing error text